File: 1942 The Number of the Smallest Unoccupied Chair/1942.py
class Solution:
    def smallestChair(self, times: List[List[int]], targetFriend: int) -> int:
        pq, chairs = [], []
        taken = {}

        for i, (arrive, leave) in enumerate(times):
            heappush(chairs, i)
            pq.append((arrive, True, i))
            pq.append((leave, False, i))

        pq.sort()

        for time, take, person in pq:
            if take:
                taken[person] = heappop(chairs)
                if person == targetFriend:
                    return taken[person]
            else:
                heappush(chairs, taken[person])
                
            

File: 1942 The Number of the Smallest Unoccupied Chair/1942.cpp
class Solution {
public:
    int smallestChair(vector<vector<int>>& times, int targetFriend) {
        vector<vector<int>> pq;
        priority_queue<int, vector<int>, greater<int>> chairs;
        unordered_map<int, int> taken;

        for (int i = 0; i<times.size(); i++){
            chairs.push(i);
            
            int arrive = times[i][0];
            int leave = times[i][1];

            pq.push_back({leave, 0, i});
            pq.push_back({arrive, 1, i});
        }

        sort(pq.begin(), pq.end());

        for (vector<int> cur: pq){
            int take = cur[1], person = cur[2];
            if (take){
                taken[person] = chairs.top(); chairs.pop();
                if (person == targetFriend) return taken[person];
            } else {   
                chairs.push(taken[person]);
            }
        }

        return 0;
    }
};
